packages/js/src/decorator/helpers/getAnnotationsHelpers.js

```diff
@@ -0,0 +1,442 @@
+		sentenceIndices.forEach( sentenceIndex => {
+			/*
+			 * The yoastmarkOffset.startOffset and yoastmarkOffset.endOffset are offsets of the <yoastmark>
+			 * relative to the start of the Mark object. The sentenceIndex is the index form the start of the
+			 * RichText until the matched Mark, so to calculate the offset from the RichText to the <yoastmark>
+			 * we need to add those offsets.
+			 *
+			 * startOffset = ( sentenceIndex ) 13 + ( yoastmarkOffset.startOffset ) 7 = 20
+			 * endOffset =   ( sentenceIndex ) 13 + ( yoastmarkOffset.endOffset ) 14  = 27
+			 *
+			 * "A cool text. A cool keyword."
+			 *      ( startOffset ) ^20   ^27 ( endOffset )
+			 */
+			const startOffset = sentenceIndex + yoastmarkOffset.startOffset;
+			let endOffset = sentenceIndex + yoastmarkOffset.endOffset;
+
+			/*
+			 * If the marks are at the beginning and the end we can use the length, which gives more
+			 * consistent results given we strip HTML tags.
+			 */
+			if ( yoastmarkOffset.startOffset === 0 && yoastmarkOffset.endOffset === mark.getOriginal().length ) {
+				endOffset = sentenceIndex + originalSentence.length;
+			}
+
+			blockOffsets.push( {
+				startOffset,
+				endOffset,
+			} );
+		} );
+	} );
+	return blockOffsets;
+}
+
+/**
+ * Creates an annotation if the given mark is position based.
+ * A helper for position-based highlighting.
+ *
+ * @param {string} clientId The client id of the block.
+ * @param {Mark}   mark The mark to apply to the content.
+ * @param {string} html The HTML of the block.
+ * @param {string} text The text of the block.
+ *
+ * @returns {Array} The annotations to apply.
+ */
+export function createAnnotationsFromPositionBasedMarks( clientId, mark, html, text ) {
+	if ( clientId === mark.getBlockClientId() ) {
+		const slicedHtml = html.slice( mark.getBlockPositionStart(), mark.getBlockPositionEnd() );
+		const slicedText = text.slice( mark.getBlockPositionStart(), mark.getBlockPositionEnd() );
+		// Check if the html and the rich text contain the same text in the specified index.
+		if ( slicedHtml === slicedText ) {
+			return [
+				{
+					startOffset: mark.getBlockPositionStart(),
+					endOffset: mark.getBlockPositionEnd(),
+				},
+			];
+		}
+		// Adjust block position start and end.
+		let startOffset = mark.getBlockPositionStart();
+		let endOffset = mark.getBlockPositionEnd();
+		// Retrieve the html from the start until the startOffset of the mark.
+		html = html.slice( 0, mark.getBlockPositionStart() );
+		// Find all html tags.
+		const foundHtmlTags = [ ...html.matchAll( htmlTagsRegex ) ];
+		/*
+		 * Loop through the found html tags backwards, and adjust the start and end offsets of the mark
+		 * by subtracting them with the length of the found html tags.
+		 */
+		for ( let i = foundHtmlTags.length - 1; i >= 0; i-- ) {
+			const currentHtmlTag = foundHtmlTags[ i ][ 0 ];
+			startOffset -= currentHtmlTag.length;
+			endOffset -= currentHtmlTag.length;
+		}
```

Comment:
This step is necessary to account for the difference in the way we "parse" the block and calculate the token position between `yoastseo `package and block annotation API.
Inside `yoastseo`, the token's position information also takes into account all the HTML tags surrounding it, in a block. However, the block annotation API applies annotations to "clean" text/html without any HTML tags. As a result, the token position information we retrieve from `yoastseo` wouldn't match that of block annotation API.